### ISSUE TITLE
Would like "fail" and "complete" directives for conversation syntax

### ISSUE THREAD

**Issue by Pointedstick** [endless-sky/endless-sky#1563]
In a conversation, "decline" will make a mission instantly be declined and jump to the "on decline" section. But there's no "fail" or "complete" directives that will instantly make the mission fail or complete and jump to the appropriate section.

You might be wondering why I'm trying to do this, and it's because I'm writing a "mission" that executes solely on a planet using conversation; by the time the player takes off, it should already be failed or completed. I'd like to fine the player in the "on fail" section and pay them in the "on complete" section but I don't see a straightforward to do this within a conversation. Basically it would look as follows, if this worked:

```
on accept
    conversation
            `You decide to gamble.`
            branch "good luck" "bad luck"
                random < 50

            label "good luck"
            `   You win some money!`
                complete

            label "bad luck"
            `   You lose some money!`
                fail

    on fail
        payment -25000

    on complete
        payment 15000
```

**Comment by KiLEdEnNis** [comment]
enabling mission branches like 1 that "completes" and one that "fails" is something like that right?
(so someone could fail a mission and influence the story like that)

**Comment by Pointedstick** [comment]
Alternatively, #1564 may be a better approach for what I'm trying to accomplish.

**Comment by endless-sky** [comment]
Why not just use "accept" and "decline," put the payment in the "on accept"
and "on decline" actions, and have "on accept" also "fail" the mission?

**Comment by KiLEdEnNis** [comment]
not sure but #1564 is too complex for that approach right?

**Comment by Pointedstick** [comment]
Correct, the conversation in #1564 is too complex for that approach: there should be no payment for declining (you don't gamble with the pirates), and multiple different payments depending on how your gambling/the conversation does.

Regardless, even if I simplified it, putting a payment in "on accept" wouldn't work because #1562 makes conversations not work in "on accept" blocks.

**Comment by endless-sky** [comment]
You could do the following:

* If you decline to gamble, "defer."
* If you choose to gamble, branch based on "random."
* If you win, show the text for winning, and then end the conversation with
"accept."
* If you lose, show the text for losing, and then "decline."
* In "on accept", give the payment for winning, then "fail".
* In "on decline", give the payment for losing.

**Comment by Pointedstick** [comment]
Looks like that _generally_ works. It's not as quite as flexible as I'd like to accommodate all the outcomes I'm envisioning for this repeatable mission:

1. You win some money (accept; "on accept" has payment and fail)
2. You lose some money (decline, "on decline" has negative payment and fail)
3. You win some money and blast off from the planet alongside two pirate ships (launch; "on accept" has payment and fail)
4. You lose some money and blast off from the planet alongside two pirate ships (flee, "on decline" has negative payment and fail)

2 and 3 work fine (except for issue #1567). With 1, the pirate ships attack you anyway, presumably because you triggered "on accept". With 4, no pirate ships attack you, presumably because you triggered "on decline"

**Comment by KiLEdEnNis** [comment]
even if it works fine, the missions need to be able to have multiple outcomes, with different payments or whatever

**Comment by Pointedstick** [comment]
I'm working on another mission that I can't satisfactorily finish without this. I want to put a conversation in the "on enter" block to make it appear as soon as you take off, and allow one outcome of the conversation to make the mission fail instantly and trigger whatever's in the "on fail" block.

I would love to add many more missions to this game, but every time I try to do anything complicated with branching conversations that can trigger different outcomes, I feel like I run into this issue or https://github.com/endless-sky/endless-sky/issues/1564, and I get stuck.

**Comment by Pointedstick** [comment]
@tehhowch a hint on how to implement this might set me in the right direction for producing a patch. :slightly_smiling_face:

**Comment by tehhowch** [comment]
As a first thought, you'd need to define new Conversation exit values, e.g. Conversation::FAIL and Conversation::COMPLETE
Then you'd need to address the issues associated with #3228 , #3206 , #3167, and #3129: conversation callbacks from Mission::Do / MissionAction::Do / NPC::Do to handle the conversation's exit result.

(to do your payment accumulation, you could use an `apply` node to keep updating a condition `"my winnings"` based on the player's result, and then reference that condition in the `on accept`/`on complete`/`on fail`.
You'll still need a source change to allow use of a payment amount that is read from a condition instead of a fixed number though, so the more direct route of #1564 may be preferred.)

**Comment by Pointedstick** [comment]
Yikes, that sounds somewhat above my abilities at this point. I find the C++ code in this game to be really challenging to work with as a novice.

**Comment by tehhowch** [comment]
I'm sure you can hack it :+1:
The existing callback is PlayerInfo::MissionCallback, so you can get ideas on how to handle the new exit values from there, in addition to how to register your new callback from looking at how that one is used.

**Comment by tehhowch** [comment]
#3376 initially went towards giving you almost everything you'd need for full support of this (Conversations called with callbacks whenever generated from a mission or NPC), but the latest refactor has dropped the use of a "secondary" callback since the only thing that callback was doing was enabling player death.
If (when?) #3376 merges, you can use the development history in the branch's commits to faithfully recreate the `PlayerInfo::BasicCallback` method and its invocations.

The last tricky part is to find the relevant mission within the player's mission list and perform its fail or complete MissionAction (`mission.Do(<blah>)`).
I'd probably save payment accumulation and extraction from a condition variable for a different PR.

**Comment by tehhowch** [comment]
Adding the BasicCallback would also enable #3167 to be handled, so perhaps I should restore that in my PR. It's a touch beyond the original scope, but the PR is essentially reworking conversation exits anyway.

Then you'd just need to add the handling mentioned above for your conversation responses to both callbacks.

**Comment by tehhowch** [comment]
BasicCallback for non-`on offer` conversations added in #3420 , but I'm not certain you're going to be able to handle them in the callback at all, because you have no nice way to reference the mission you want to fail or complete.
The `on offer` conversations side-step this issue by always having the mission being offered at the front of the list - don't need an index or iterator to the mission when you know it's the first one! For your purposes though, the desired mission to fail or complete can be at any position within the player's `missions` list.
So to do this well, you'd need to (as I do in #3376) extend the handling of the response in ConversationPanel::Exit (e.g.
```
if ( node is fail or complete )
    missionlist = player.Missions
    for ( mission in missionlist )
        if (is this mission the one i want to fail/complete )
            mission->Fail | mission->Complete | mission->Do(mission::FAIL) | mission->Do(mission::COMPLETE) (something like this)
```

**Comment by Amazinite** [comment]
I spoke with Pointedstick about this, and any conceivable use of this enhancement seemed to be better solved by #1564, so I'm closing this issue.
